Crop option for thumbnail types

mnakalay opened the pull request:
Takes care of #3909 

Added a crop option to the thumbnail type system

It should be backward compatible

It also takes care of modifications during C5 install.

Check this video: https://www.youtube.com/watch?v=-Xn9z_WFHZ4

Mnkras commented on concrete/controllers/single_page/dashboard/system/files/thumbnails.php:
```diff
@@ -12,16 +12,40 @@ public function view()
+    public function get_sizing_values()
+    {
+        $sizingModes = array(
+            Type::RESIZE_PROPORTIONAL => 'Resize Proportionally',
```
Are these constants or should they have t()?

mnakalay replied:
Oups sorry they definitely should have t()

Mnkras commented on concrete/src/Entity/File/Version.php:
```diff
@@ -1010,6 +1010,7 @@ public function getImagineImage()
      */
     public function rescanThumbnails()
     {
+        // [NOUR Look here and in generateThumbnail()]
```
?

mnakalay replied:
Oups again, leftover. I'll delete it

Mnkras commented on concrete/src/File/Image/Thumbnail/Type/Version.php:
```diff
@@ -14,14 +14,16 @@ class Version
     protected $isDoubledVersion;
 
-    public function __construct($directoryName, $handle, $name, $width, $height, $isDoubledVersion = false)
+    public function __construct($directoryName, $handle, $name, $sizingMode, $width, $height, $isDoubledVersion = false)
```
This is not BC, this will break stuff. Specifically the CustomThumbnail Type

mnakalay replied:
ok I'll check.
What do you mean by BC?

mnakalay replied:
I just figured out BC. I suppose you mean Backward Compatible? Duh!!! Sorry, not a native English speaker. My brain has to work hard to get acronyms...

Anyway, I think my last modification deals with that

Mnkras replied:
This still breaks BC because if anyone has extended this, their code will break, if anything this parameter should be added as an optional parameter at the end of the function signature as PHP will ignore extra parameters

Mnkras replied:
This option NEEDS to be optional, it cannot be required.

Mnkras commented:
This also needs a migration to be written.

mnakalay commented:
I have to admit Doctrine, migration, all that, you have to be more explicit. My understanding is the entity is modified and will add the necessary fields to the table.
Is there anything else to do?

mnakalay commented:
@Mnkras I removed the comment, added the two t() functions and modified CustomThumbnail to work with the new system without having t modify the other stuff that use it.

Mnkras commented on concrete/controllers/single_page/dashboard/system/files/thumbnails.php:
```diff
@@ -142,16 +185,22 @@ public function do_add()
             $height = intval($request->request->get('ftTypeHeight'));
+            $width = intval($request->request->get('ftTypeWidth'));
             if ($height > 0) {
                 $type->setHeight($request->request->get('ftTypeHeight'));
```
Can you change this to `$height`?

Mnkras commented on concrete/controllers/single_page/dashboard/system/files/thumbnails.php:
```diff
@@ -142,16 +185,22 @@ public function do_add()
             }
-            $type->setWidth($request->request->get('ftTypeWidth'));
+            if ($width > 0) {
+                $type->setWidth($request->request->get('ftTypeWidth'));
```
and `$width` here

Mnkras commented:
Once one of the core team takes a look at this we can write the migration, it should be pretty straight forward.

mnakalay commented:
Hi again @Mnkras. I modified it again, the sizing mode parameter comes last now and has a default value and I made sure everywhere the class was instantiated, proper parameters were in place.

mlocati commented on concrete/controllers/single_page/dashboard/system/files/thumbnails.php:
```diff
@@ -113,15 +149,22 @@ public function update()
-            $type->setWidth($request->request->get('ftTypeWidth'));
+            
             $height = intval($request->request->get('ftTypeHeight'));
+            $width = intval($request->request->get('ftTypeWidth'));
```
Just a (really) minor performance note: `(int) $value` is much faster than `intval($value)`

mnakalay replied:
Hi @mlocati. I have no problem with using one of the other but I used what was already there. The core team had used intval so I kept it that way. Should I change it?

mlocati replied:
Why not... :wink:

mlocati commented on concrete/controllers/single_page/dashboard/system/files/thumbnails.php:
```diff
@@ -113,15 +149,22 @@ public function update()
                 $type->setHeight(null);
             }
+            if ($width > 0) {
+                $type->setWidth($request->request->get('ftTypeWidth'));
```
What about using `$width`?

mlocati commented on concrete/controllers/single_page/dashboard/system/files/thumbnails.php:
```diff
@@ -12,16 +12,40 @@ public function view()
+        );
+        $this->set('sizingModes', $sizingModes);
+
+        $sizingModeHelp = array(
```
If you run `php-cs-fixer` we'll have the (much nicer) short array syntax

mnakalay replied:
I did but it didn't change the array.
should I change it manually everywhere I added arrays?

mlocati replied:
I think you missed the `--config` parameter:
```php-cs-fixer fix --config=<path-to-c5-root>/.php_cs.dist thumbnails.php```

I configured my IDE to have a button that, when clicked, calls `php-cs-fixer fix --config=<path-to-c5-root>/.php_cs.dist` on the currently selected file

mlocati commented on concrete/single_pages/dashboard/system/files/thumbnails.php:
```diff
@@ -126,8 +137,9 @@
         <td><a href="<?=$view->action('edit', $type->getID())?>"><?=$type->getHandle()?></a></td>
         <td><?=$type->getDisplayName()?></td>
-        <td><?=$type->getWidth()?></td>
+        <td><?=($type->getWidth()) ? $type->getWidth() : '<span class="text-muted">' . t('Automatic') . '</span>' ?></td>
```
There's no need for the extra parenthesis around `$type->getWidth()`

mnakalay replied:
@mlocati again agree but I just copied the line that was already there for height. Usually, when the core team does something a certain way I leave it like that in case it's the way they like it. DO you think I should remove the parenthesis?

mlocati replied:
Again... why not? :wink:

mnakalay commented:
@mlocati and @Mnkras you both made the same comment asking what about using $height and $width that I didn't understand.

Do you mean using if ($width) and if ($height) instead of if ($width > 0) and if ($height >0)?

If that's it, again, I kept what the core team had in place, should I modify it?

mlocati commented:
I meant code like this
```php
$height = intval($request->request->get('ftTypeHeight'));
// [...]
if ($height > 0) {
    $type->setHeight($request->request->get('ftTypeHeight'));
}
```
can be rewritten as
```php
$height = intval($request->request->get('ftTypeHeight'));
// [...]
if ($height > 0) {
    $type->setHeight($height);
}
```
and I'm pretty sure @Mnkras meant the same

mlocati commented:
> Usually, when the core team does something a certain way I leave it like that in case it's the way they like it

Here's what I think:
1. the previous code may me literally ten years old
2. the previous code can be written by contributors, not only by the core team

So, when I see some code that can be improved and/or optimized, I do it

mnakalay commented:
@mlocati this is embarrassing. I kept staring at those lines and couldn't understand for the life of me what was wrong until you pointed it out. I guess sometimes when you're too close to something it's all a blur. It's corrected now.
I'm going to also correct the other little things.

mlocati commented:
> I guess sometimes when you're too close to something it's all a blur.

I totally agree: that's why github reviews are so helpful :wink:

mnakalay commented:
ok. Should be good now...

aembler commented:
Uh oh, I think I broke this one with the merging of #5424

mnakalay commented:
ok @aembler and @Mnkras I think conflicts are resolved on this one

mnakalay commented:
I don't know why Travis is failing. It seems to be a generalized problem

aembler commented:
Thanks! I added a migration
